File: Medium/next-permutation.cpp
//https://leetcode.com/problems/next-permutation/
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
     int n = nums.size(); // Size of the given array
        int i = n - 2; // Start from the second last element

        // Find the first element that is not in a non-increasing sequence from the end
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            --i;
        }

        if (i >= 0) { // If such an element is found
            // Find the smallest element greater than nums[i] to the right of it
            for (int j = n - 1; j > i; --j) {
                if (nums[j] > nums[i]) {
                    std::swap(nums[i], nums[j]); // Swap them
                    break;
                }
            }
        }

        // Reverse the sequence from nums[i + 1] to the end of the array
        // This is to get the smallest possible permutation after i
        std::reverse(nums.begin() + i + 1, nums.end());
    }
};

File: Medium/Next Permutation.java
//https://leetcode.com/problems/next-permutation/description/
import java.util.Arrays;

class Solution {
    public void nextPermutation(int[] nums) {
        // Initialize index to -1, which will be used to store the position
        // where the next greater permutation needs to be found
        int index = -1;
        int n = nums.length;

        // Find the first index from the end where the current element is less than the next element
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                index = i;
                break;
            }
        }

        // If no such index is found, reverse the entire array to get the smallest permutation
        if (index == -1) {
            reverse(nums, 0, n - 1);
            return;
        }

        // Find the first element from the end that is greater than the element at the found index
        for (int i = n - 1; i > index; i--) {
            if (nums[i] > nums[index]) {
                swap(nums, i, index);
                break;
            }
        }

        // Reverse the elements after the found index to get the next permutation
        reverse(nums, index + 1, n - 1);
    }

    // Helper method to reverse elements in the array
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }

    // Helper method to swap elements in the array
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
